code/game/atoms_movable.dm

```diff
@@ -399,15 +423,24 @@
 	loc = newloc
 
 	. = TRUE
-	oldloc.Exited(src, newloc)
+
+	if(old_locs)
+		for(var/atom/exited_loc as anything in (old_locs - new_locs))
+			exited_loc.Exited(src, direction)
+	else
+		oldloc.Exited(src, direction)
 	if(oldarea != newarea)
-		oldarea.Exited(src, newloc)
+		oldarea.Exited(src, direction)
 
-	newloc.Entered(src, oldloc)
+	if(new_locs)
+		for(var/atom/entered_loc as anything in (new_locs - old_locs))
+			entered_loc.Entered(src, direction)
+	else
+		newloc.Entered(src, direction)
 	if(oldarea != newarea)
-		newarea.Entered(src, oldloc)
+		newarea.Entered(src, direction)
 
```

Comment:
Does this add any overhead to moving stuff? do you mind taking a profile of it?

Reply:
I was concerned as well with perf, so I did my best to optimize it.

If the movable is a single-tile object then all this adds for overhead is a few if checks.
Only if the movable is multi-tiled it will have to do some extra work, and even then, this shouldn't be costly at all.

Added some comments, check them out. The locs list is not even populated if it's not a multi-tile object, so the logic remains the same for most things.